### PR TITLE
show ggml modelinfo through the show api

### PR DESCRIPTION
**Pull request by pdevine**
This change exposes the GGML KVs and tensor data to make it easier to introspect a model.

### CODE REVIEW

File: server/routes.go [hunk 1]
```diff
@@ -721,12 +721,61 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	}
+
+	kvMap := make(map[string]any)
+
+	for _, k := range keys {
+		val := kv[k]
+
+		switch v := val.(type) {
+		case []interface{}:
+			if len(v) > 5 {
```
Comment (BruceMacD) [inline server/routes.go:759]: Why skip values longer than length 5?

**Review by BruceMacD — COMMENTED** [review]
Is this being used in conversion? Wondering what the reason is for surfacing this.

File: server/routes.go [hunk 2]
```diff
@@ -721,12 +721,61 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
 	return resp, nil
 }
 
+func getGGMLData(model *Model) ([]byte, error) {
+	f, err := os.Open(model.ModelPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ggml, _, err := llm.DecodeGGML(f)
```
Comment (mxyng) [inline server/routes.go:741]: nice be useful to return the total size in model info

File: server/routes.go [hunk 3]
```diff
@@ -721,12 +721,61 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	kv := ggml.KV()
+	var keys []string
```
Comment (mxyng) [inline server/routes.go:747]: keys isn't being sorted so I don't see a reason to create this slice

File: api/types.go [hunk 4]
```diff
@@ -207,6 +207,7 @@ type ShowResponse struct {
 	System     string       `json:"system,omitempty"`
 	Details    ModelDetails `json:"details,omitempty"`
 	Messages   []Message    `json:"messages,omitempty"`
+	ModelInfo  string       `json:"modelinfo,omitempty"`
```
Comment (mxyng) [inline api/types.go:210]: Since this is actually JSON, `json.RawMessage` would be more appropriate

File: server/routes.go [hunk 5]
```diff
@@ -721,12 +721,61 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+
+	kvMap := make(map[string]any)
+
+	for _, k := range keys {
+		val := kv[k]
+
+		switch v := val.(type) {
+		case []interface{}:
+			if len(v) > 5 {
+				kvMap[k] = []string{}
```
Comment (mxyng) [inline server/routes.go:760]: It's confusing to see an empty JSON list as the value. This should have something that indicates the type and size of the value, something like `(list: 280K items)`

File: server/routes.go [hunk 6]
```diff
@@ -721,12 +721,61 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+				continue
+			}
+		}
+		kvMap[k] = val
```
Comment (mxyng) [inline server/routes.go:764]: nit: this can go into default to omit the continue in the other case

**Pull request closed without merging**